File: LCAofaBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        # TC - O(N) not log N because what if skewed tree is given
        # SC - O(1) if iterative is implemented O(N) for recursive stack (implicit or explicit)
        
        #base case
        if root == None:
            return 0
        
        pval = p.val
        qval = q.val
        
        node = root
        
        while(node):
            
            value = node.val
            
            #both nodes on left side
            if(pval<value and qval<value):
                node=node.left
            
            # both nodes on right side
            elif (pval>value and qval>value):
                node = node.right
             
            # node is the solution
            else:
                return node
            
        return None
    
        '''
        #TC - O(N)
        #SC - O(N) recursive stack
        # recursive solution
        
        val = root.val
        
        pval = p.val
        qval = q.val
        
        if (pval>val and qval>val):
            return lowestCommonAncestor(root.right,p,q)
            
        elif (pval<val and qval<val):
            return lowestCommonAncestor(root.left,p,q)
            
        else:
            return root
        
        '''

File: LCA of a BST.java

 // Time Complexity : O(h) 
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
// Your code here along with comments explaining your approach
// We will look for a root node which has both p and q on two different sides. If they are on the same side, then we keep moving forward until 
// we find a node which lies between the two numbers. 

 class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {

        //check if both the elements are on the left sides
        if(root.val>p.val && root.val>q.val)
        {
            return lowestCommonAncestor(root.left,p,q);
        }
        //check if they are on the right sides
        else if(root.val<p.val && root.val<q.val)
        {
            return lowestCommonAncestor(root.right,p,q);
        }
        //else if they are on two different sides or p is the child of q or vice versa, then the root is the ancestor
        else
        {
            return root;
        }
        
    }
}